File: v2/compression.ts
/**
 * Compression Module
 *
 * Shared LZ-string compression/decompression for rule logic.
 * Extracted from rule-logic-lib.ts for reuse in v2.
 */

import LZString from 'lz-string';

// =============================================================================
// COMPRESSION
// =============================================================================

/**
 * Compress rule logic to Foundry's expected format
 *
 * @param logic - The rule logic object
 * @returns JSON string containing { compressedValue, type: 'compressedValue' }
 */
export function compress(logic: unknown): string {
  const json = JSON.stringify(logic);
  const compressed = LZString.compressToEncodedURIComponent(json);
  return JSON.stringify({ compressedValue: compressed, type: 'compressedValue' });
}

/**
 * Decompress rule logic from Foundry's format
 *
 * @param compressedWrapper - JSON string containing compressedValue
 * @returns Parsed rule logic object
 */
export function decompress(compressedWrapper: string): unknown {
  const wrapper = JSON.parse(compressedWrapper);
  const json = LZString.decompressFromEncodedURIComponent(wrapper.compressedValue);
  if (!json) {
    throw new Error('Failed to decompress - invalid compressed value');
  }
  return JSON.parse(json);
}

/**
 * Get just the compressed value string (without wrapper)
 */
export function getCompressedValue(compressedWrapper: string): string {
  const wrapper = JSON.parse(compressedWrapper);
  return wrapper.compressedValue;
}

/**
 * Create compressed wrapper from raw compressed value
 */
export function wrapCompressedValue(compressedValue: string): string {
  return JSON.stringify({ compressedValue, type: 'compressedValue' });
}

/**
 * Get compression stats
 */
export function getCompressionStats(logic: unknown): {
  originalSize: number;
  compressedSize: number;
  ratio: number;
} {
  const original = JSON.stringify(logic);
  const compressed = LZString.compressToEncodedURIComponent(original);

  return {
    originalSize: original.length,
    compressedSize: compressed.length,
    ratio: compressed.length / original.length
  };
}

File: src/foundry_rules/compression.py
"""
Compression Module

LZ-string compression/decompression for rule logic.
"""

import json
from typing import Any

import lzstring


# Create LZ-string compressor instance
_lz = lzstring.LZString()


def compress(logic: Any) -> str:
    """
    Compress rule logic to Foundry's expected format.

    Args:
        logic: The rule logic object (dict)

    Returns:
        JSON string containing { compressedValue, type: 'compressedValue' }
    """
    json_str = json.dumps(logic)
    compressed = _lz.compressToEncodedURIComponent(json_str)
    wrapper = {"compressedValue": compressed, "type": "compressedValue"}
    return json.dumps(wrapper)


def decompress(compressed_wrapper: str) -> Any:
    """
    Decompress rule logic from Foundry's format.

    Args:
        compressed_wrapper: JSON string containing compressedValue

    Returns:
        Parsed rule logic object (dict)

    Raises:
        ValueError: If decompression fails
    """
    wrapper = json.loads(compressed_wrapper)
    compressed_value = wrapper.get("compressedValue")

    if not compressed_value:
        raise ValueError("Missing compressedValue in wrapper")

    json_str = _lz.decompressFromEncodedURIComponent(compressed_value)

    if not json_str:
        raise ValueError("Failed to decompress - invalid compressed value")

    return json.loads(json_str)
